retain_graph=True in compiled_autograd

voznesenskym opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110367

Adds support for retain_graph=True - known as keep_graph_ internally in the autograd engine. 

cc @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

voznesenskym committed:
keep_graph in compiled_autograd

keep_graph in compiled_autograd

keep_graph in compiled_autograd

[ghstack-poisoned]

voznesenskym commented on test/inductor/test_compiled_autograd.py:
```diff
@@ -389,121 +426,114 @@ def wrapped(self: EagerAutogradTests):
-    "test_hook_closure_cycle_use_custom_function_False_use_tensor_hook_True",  # compiled_autograd does not support keep_graph
-    "test_hook_closure_cycle_use_custom_function_True_use_tensor_hook_False",  # compiled_autograd does not support keep_graph
-    "test_hook_closure_cycle_use_custom_function_True_use_tensor_hook_True",  # compiled_autograd does not support keep_graph
+    "test_accumulate_grad",  # RuntimeError: compiled_autograd does not support create_graph
```
10 tests passed now, but regenerated all to get new errors.

jansel commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -30,6 +30,15 @@ def maybe_clone(x):
     return x
 
 
+def keep_graph(graph, keep_graph):
+    graph.keep_graph = keep_graph
+    return graph
+
+
+def should_keep_graph(graph):
+    return graph.keep_graph
```
Why does this need to be in python?

voznesenskym replied:
Setting breakpoints / debugging is easier ;)

I can move entirely to C++ setting

jansel reviewed (changes requested):
Looks like tests are failing

voznesenskym committed:
Update on "keep_graph in compiled_autograd"


keep_graph in compiled_autograd

keep_graph in compiled_autograd

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

voznesenskym commented:
> Looks like tests are failing

Most of the actual tests seem to pass, I think its a static analysis catching (correctly) a falsey condition leading to uninitialized ptr access. 

Fixed.

voznesenskym committed:
Update on "keep_graph in compiled_autograd"


keep_graph in compiled_autograd

keep_graph in compiled_autograd

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

jansel commented on torch/csrc/dynamo/python_compiled_autograd.cpp:
```diff
@@ -369,7 +402,9 @@ variable_list compiled_autograd(
-  if (!cache->check_dynamic_sizes(compiler_call)) {
+  // TODO(voz): Replace with a general cache check fn
+  if (!cache->check_dynamic_sizes(compiler_call) ||
+      !cache->check_keep_graph(compiler_call)) {
```
I think instead of doing this, we should just encode keep_graph in the cache key.

We could do something right after:
https://github.com/pytorch/pytorch/blob/4bbe41bc05f48ec5dc61559887ced9a983019630/torch/csrc/dynamo/python_compiled_autograd.cpp#L361C22-L361C40
like:
```
if(keep_graph) {
   cache = cache->lookup(... fixed key to mark usage of keep_graph ...)
}
```

ezyang replied:
Given the caching discussion, I believe that this diff can be deleted entirely

jansel commented on torch/csrc/dynamo/compiled_autograd.h:
```diff
@@ -174,6 +174,7 @@ struct AutogradCompilerCall {
   std::vector<c10::SafePyObject> hooks;
   NodeCalls node_calls;
   SizeInput::DynType default_dyn_type = SizeInput::STATIC;
+  bool keep_graph = false;
```
I don't like exposing this in python, this is not something that eager does and is much slower to read/write than a C++ variable.

ezyang replied:
I don't think you need this anyway?

jansel commented:
On second thought, do we even need to recompile when keep_graph changes?  I don't think it will affect the generated FX graph? Maybe we can just not release things?

voznesenskym commented:
> On second thought, do we even need to recompile when keep_graph changes? I don't think it will affect the generated FX graph? Maybe we can just not release things?

It's a good question. While it won't affect the generated FX graph, I was worried about preserving callsite semantics. 

```
.backward(retain_graph=True)
.backward(retain_graph=False)
```
If we cached the first graph, and do not recompile, we ignored a user directive to clean up the graph/release memory. 

It should be sound from a correctness and fx graph perspective tho. 

Lmk, if you decide this is fine, the other comments on the PR become moot (as they pertain to the caching)

voznesenskym commented:
> > On second thought, do we even need to recompile when keep_graph changes? I don't think it will affect the generated FX graph? Maybe we can just not release things?
> 
> It's a good question. While it won't affect the generated FX graph, I was worried about preserving callsite semantics.
> 
> ```
> .backward(retain_graph=True)
> .backward(retain_graph=False)
> ```
> 
> If we cached the first graph, and do not recompile, we ignored a user directive to clean up the graph/release memory.
> 
> It should be sound from a correctness and fx graph perspective tho.
> 
> Lmk, if you decide this is fine, the other comments on the PR become moot (as they pertain to the caching)

Actually, I am wrong here:

```
if (!graph_task.keep_graph_) {
    for (auto& call : calls) {
      call->node->release_variables();
    }
  }
```
Is outside the compile region. I think it should be safe to just not consider this w/r/t caching.

jansel commented:
Yeah, I think we can just ignore this for caching, but please at a test that will have a cached version with the opposite flag set.

ezyang commented:
Need more pr desc

ezyang commented:
Needs PR desc?

Also, why is this called `keep_graph`, isn't the preexisting term `retain_graph`?

ezyang commented:
Usually you only need `retain_graph` for double backwards like situations, are we running into this?

ezyang commented:
A helpful addition to the PR description would be to describe, concretely, what the preexisting `retain_graph` functionality in autograd engine does, and then describe how that functionality is replicated in compiled autograd

ezyang commented on torch/csrc/dynamo/python_compiled_autograd.cpp:
```diff
@@ -461,16 +496,20 @@ variable_list compiled_autograd(
-    cache->compiled_fn = check(call_end_capture(py_compiler, state.outputs));
+    PyObject* graph = check(call_end_capture(py_compiler, state.outputs));
+    cache->compiled_fn =
+        check(call_keep_graph(check(graph), graph_task.keep_graph_));
```
this can be removed

albanD commented:
> Usually you only need retain_graph for double backwards like situations, are we running into this?

It is enabled when `create_graph=True` and thus in this situation it is yes.
It is also used when you need to backward a given model multiple times without changing it (like in GAN use cases).

voznesenskym committed:
Update on "retain_graph=True in compiled_autograd"


Adds support for retain_graph=True - known as keep_graph_ internally in the autograd engine. 

cc penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot merge

voznesenskym commented:
@pytorchbot merge -f "Flaky CI"

voznesenskym commented:
<img width="489" alt="image" src="https://github.com/pytorch/pytorch/assets/4755252/97d5dc7e-8054-4df1-9a41-adbfc2cc09f4">
